From 172cd4532f491afb462d9da9ba7ea92a709a9dc5 Mon Sep 17 00:00:00 2001
From: kongweibin <kongweibin2@huawei.com>
Date: Thu, 6 Jul 2023 15:32:53 +0800
Subject: [PATCH 06/36] [Backport] bpf: Allow bpf_get_netns_cookie in
 BPF_PROG_TYPE_SOCK_OPS

mainline inclusion
from mainline-v5.15-rc1
commit 6cf1770d63dd2d0d0d4048e7b3ee360336c072d9
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I7JF40

Reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=6cf1770d63dd2d0d0d4048e7b3ee360336c072d9

--------------------------------

We'd like to be able to identify netns from sockops hooks to
accelerate local process communication form different netns.

Signed-off-by: Xu Liu <liuxu623@gmail.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Acked-by: Song Liu <songliubraving@fb.com>
Link: https://lore.kernel.org/bpf/20210818105820.91894-2-liuxu623@gmail.com
Signed-off-by: Kong Weibin <kongweibin2@huawei.com>
---
 net/core/filter.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/net/core/filter.c b/net/core/filter.c
index 75db9a5b3cff..4f16631dded6 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -4685,6 +4685,18 @@ static const struct bpf_func_proto bpf_get_netns_cookie_sock_addr_proto = {
 	.arg1_type	= ARG_PTR_TO_CTX_OR_NULL,
 };
 
+BPF_CALL_1(bpf_get_netns_cookie_sock_ops, struct bpf_sock_ops_kern *, ctx)
+{
+	return __bpf_get_netns_cookie(ctx ? ctx->sk : NULL);
+}
+
+static const struct bpf_func_proto bpf_get_netns_cookie_sock_ops_proto = {
+	.func		= bpf_get_netns_cookie_sock_ops,
+	.gpl_only	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_PTR_TO_CTX_OR_NULL,
+};
+
 BPF_CALL_1(bpf_get_socket_uid, struct sk_buff *, skb)
 {
 	struct sock *sk = sk_to_full_sk(skb->sk);
@@ -7481,6 +7493,8 @@ sock_ops_func_proto(enum bpf_func_id func_id, const struct bpf_prog *prog)
 		return &bpf_sk_storage_get_proto;
 	case BPF_FUNC_sk_storage_delete:
 		return &bpf_sk_storage_delete_proto;
+	case BPF_FUNC_get_netns_cookie:
+		return &bpf_get_netns_cookie_sock_ops_proto;
 #ifdef CONFIG_EULER_SOCKETMAP
 	case BPF_FUNC_get_sockops_uid_gid:
 		return &bpf_get_sockops_uid_gid_proto;
-- 
2.17.1

